### PR TITLE
Allow writing to read-only file when creating

### PR DESCRIPTION
**Pull request by secfree**
### What changes are proposed in this pull request?

Allow writing to read-only file when creating

### Why are the changes needed?

Fix https://github.com/Alluxio/alluxio/issues/15808

### Does this PR introduce any user facing changes?

No

### CODE REVIEW

**Comment by yyongycy** [comment]
does this change the behavior of hdfs?

**Comment by secfree** [comment]
> does this change the behavior of hdfs?

No. 
For HDFS, namenode only checks if the client has "write" permission while calling "createFile". If yes, namenode will give the client a lease and it only checks the lease while calling "addBlock".

**Comment by secfree** [comment]
> 1. what is the behavior on the raw hdfs?

Below is a testing on hdfs with webhdfs.

"write" to a read-only file while creating succeed.

```
$ curl -L -i -X PUT -T test-write "http://x.x.x.x:50070/webhdfs/v1/user/test/test-write?op=CREATE&permission=0400"
```

As the audit log of hdfs shows

```
2022-07-07 20:25:35,955 INFO FSNamesystem.audit: allowed=true   ugi=... (auth:SIMPLE)  ip=/...       cmd=create      src=/user/test/test-write     dst=null        perm=test:test:r--------   proto=rpc
2022-07-07 20:25:36,550 INFO FSNamesystem.audit: allowed=true   ugi=... (auth:SIMPLE)  ip=/...       cmd=complete    src=/user/test/test-write     dst=null        perm=null       proto=rpc
```

"append" to the file failed failed by AccessControlException, which is expected

```
$ curl -L -i -X POST -T test-write "http://x.x.x.x:50070/webhdfs/v1/user/test/test-write?op=APPEND"

{"RemoteException":{"exception":"AccessControlException","javaClassName":"org.apache.hadoop.security.AccessControlException","message":"Permission denied: ...
```

**Comment by secfree** [comment]
> 2\. how does alluxio hdfs api behave?

Tested with alluxio's "Hadoop-Compatible Java Client" mentioned here: https://docs.alluxio.io/os/user/stable/en/api/FS-API.html#hadoop-compatible-java-client

Below is the code (Scala code, which can use Java libraries directly and execute in spark-shell directly)

```scala
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.{FileStatus, FileSystem, Path}
import org.apache.hadoop.fs.permission.FsPermission
val conf = new Configuration()
conf.set("fs.alluxio.impl", "alluxio.hadoop.FileSystem");
conf.set("fs.AbstractFileSystem.alluxio.impl", "alluxio.hadoop.AlluxioFileSystem");
val fs = FileSystem.get(conf)
val path = new Path("alluxio://x.x.x.x:19998/user/test/file")
val file = fs.createFile(path).permission(new FsPermission("0400")).recursive().build()
file.write("hello".getBytes)
file.close()
```

The code
- Works well if using a hdfs path
- Encountered "Permission denied" for alluxio before applying this PR
- Works well for alluxio after applying this PR

**Comment by yyongycy** [comment]
> > 1. what is the behavior on the raw hdfs?
> 
> Below is a testing on hdfs with webhdfs.
> 
> "write" to a read-only file while creating succeed.
> 
> ```
> $ curl -L -i -X PUT -T test-write "http://x.x.x.x:50070/webhdfs/v1/user/test/test-write?op=CREATE&permission=0400"
> ```
> 
> As the audit log of hdfs shows
> 
> ```
> 2022-07-07 20:25:35,955 INFO FSNamesystem.audit: allowed=true   ugi=... (auth:SIMPLE)  ip=/...       cmd=create      src=/user/test/test-write     dst=null        perm=test:test:r--------   proto=rpc
> 2022-07-07 20:25:36,550 INFO FSNamesystem.audit: allowed=true   ugi=... (auth:SIMPLE)  ip=/...       cmd=complete    src=/user/test/test-write     dst=null        perm=null       proto=rpc
> ```
> 
> "append" to the file failed failed by AccessControlException, which is expected
> 
> ```
> $ curl -L -i -X POST -T test-write "http://x.x.x.x:50070/webhdfs/v1/user/test/test-write?op=APPEND"
> 
> {"RemoteException":{"exception":"AccessControlException","javaClassName":"org.apache.hadoop.security.AccessControlException","message":"Permission denied: ...
> ```

Could you please try with the code to see if a file is created with 400, and contents could be appended with the same fd?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 1]
```diff
@@ -1771,11 +1773,13 @@ public long getNewBlockIdForFile(AlluxioURI path) throws FileDoesNotExistExcepti
...
          FileSystemMasterAuditContext auditContext =
             createAuditContext("getNewBlockIdForFile", path, null, inodePath.getInodeOrNull())) {
-      try {
-        mPermissionChecker.checkPermission(Mode.Bits.WRITE, inodePath);
-      } catch (AccessControlException e) {
-        auditContext.setAllowed(false);
-        throw e;
+      if (!skipFilePermissionCheck(inodePath)) {
+        try {
+          mPermissionChecker.checkPermission(Mode.Bits.WRITE, inodePath);
```
Comment (yyongycy) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:1778]: This permission check could do other checks, it is not safe to skip here.

Comment (yyongycy) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:1778]: a possible solution is to escalate the permission temporarily, then revoke the permission after this check is done.

Comment (yyongycy) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:1778]: eg: escalate to 600 before the check if it is a newly created, then degrade to 400.

**Comment by secfree** [comment]
> Could you please try with the code to see if a file is created with 400, and contents could be appended with the same fd?

As discussed, there is a file created with 400. "append" is rejected by AccessControlException.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 2]
```diff
@@ -1787,6 +1795,22 @@ public long getNewBlockIdForFile(AlluxioURI path) throws FileDoesNotExistExcepti
     }
   }
 
+  /**
+   * In order to allow writing to read-only files when creating,
+   * we need to skip write permission check for files sometimes.
+   */
+  private boolean skipFileWritePermissionCheck(LockedInodePath inodePath)
```
Comment (yyongycy) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:1802]: how to make sure this function is called only for a file ?  what if the dir is passed in?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 3]
```diff
@@ -1771,8 +1775,12 @@ public long getNewBlockIdForFile(AlluxioURI path) throws FileDoesNotExistExcepti
          LockedInodePath inodePath = mInodeTree.lockFullInodePath(path, LockPattern.WRITE_INODE);
          FileSystemMasterAuditContext auditContext =
             createAuditContext("getNewBlockIdForFile", path, null, inodePath.getInodeOrNull())) {
+      Mode.Bits permissionNeed = Mode.Bits.WRITE;
+      if (skipFileWritePermissionCheck(inodePath)) {
+        permissionNeed = Mode.Bits.NONE;
```
Comment (yyongycy) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:1780]: Add a comment to explain why no permission is needed.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 1 above)
Comment (secfree) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: > This permission check could do other checks, it is not safe to skip here.

Yeah, skipping all checks is not a good approach.

For escalation, I searched keywords "escalate" and "escalation" in the source code, and it seems that no places using this mechanism currently.

To implement escalation, I had two ideas
1. Add "write" to the file's permission and change it back after checking. This is quite dirty and dangerous.
2. Add a new parameter for related methods to store the escalated permission. I checked and found it needs to change a lot of methods in several classes.

So I choose to decrease the permission to be checked.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 2 above)
Comment (secfree) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:1808]: Added checking `inodePath.getInode().isFile()` in the method

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 3 above)
Comment (secfree) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:1786]: Added

**Review by yyongycy — APPROVED** [review]
lgtm, thanks for fixing it.

**Review by LuQQiu — APPROVED** [review]
LGTM, thanks!

**Comment by yyongycy** [comment]
alluxio-bot, merge this please

**Pull request merged**